src/core/Carousel/Carousel.tsx

```diff
@@ -0,0 +1,298 @@
+  const [slideCount, setSlideCount] = React.useState(0);
+
+  const [keysPressed, setKeysPressed] = React.useState<Record<string, boolean>>(
+    {},
+  );
+
+  const handleKeyDown = (event: React.KeyboardEvent) => {
+    if (event.altKey || event.ctrlKey || event.metaKey) {
+      return;
+    }
+
+    if (event.key === 'ArrowLeft' || event.key === 'ArrowRight') {
+      setKeysPressed((old) => ({ ...old, [event.key]: true }));
+    }
+
+    switch (event.key) {
+      case 'ArrowLeft': {
+        setCurrentIndex((old) => (slideCount + old - 1) % slideCount);
+        break;
+      }
+      case 'ArrowRight': {
+        setCurrentIndex((old) => (slideCount + old + 1) % slideCount);
+        break;
+      }
+    }
+  };
+
+  const handleKeyUp = (event: React.KeyboardEvent) => {
+    if (event.key === 'ArrowLeft' || event.key === 'ArrowRight') {
+      setKeysPressed((old) => ({ ...old, [event.key]: false }));
+    }
+  };
+
+  return (
+    <div
+      className='iui-carousel'
+      aria-roledescription='carousel'
+      tabIndex={0}
+      onKeyDown={handleKeyDown}
+      onKeyUp={handleKeyUp}
+      {...rest}
+    >
+      <CarouselContext.Provider
+        value={{
+          currentIndex,
+          setCurrentIndex,
+          slideCount,
+          setSlideCount,
+          keysPressed,
+        }}
+      >
+        {children}
+      </CarouselContext.Provider>
+    </div>
+  );
+};
+
+const CarouselSlider = (props: React.ComponentPropsWithoutRef<'ol'>) => {
+  const context = React.useContext(CarouselContext);
+
+  if (!context) {
+    throw new Error('CarouselSlider must be used within Carousel');
+  }
+
+  const { currentIndex, setSlideCount } = context;
+
+  React.useLayoutEffect(() => {
+    setSlideCount(React.Children.count(props.children));
+  }, [props.children, setSlideCount]);
+
+  const items = React.useMemo(
+    () =>
+      React.Children.map(props.children, (child, index) =>
+        React.isValidElement(child)
+          ? React.cloneElement(child, { 'aria-hidden': index !== currentIndex })
+          : child,
+      ) ?? [],
+    [props.children, currentIndex],
+  );
+
+  return (
+    <ol
+      {...props}
+      className={cx('iui-carousel-slider', props.className)}
+      style={{
+        // height: '100%',
+        transform: `translateX(-${currentIndex * 100}%)`,
+        ...props.style,
+      }}
+    >
+      {items}
+    </ol>
+  );
+};
+
+const CarouselSlide = (props: React.ComponentPropsWithoutRef<'li'>) => {
+  const { className, children, ...rest } = props;
+
+  return (
+    <li
+      className={cx('iui-carousel-slider-item', className)}
+      role='tabpanel'
+      aria-roledescription='slide'
+      {...rest}
+    >
+      {children}
+    </li>
+  );
+};
+
+const CarouselNavigation = (props: React.ComponentPropsWithoutRef<'nav'>) => {
+  const { className, children, ...rest } = props;
+
+  const context = React.useContext(CarouselContext);
+  if (!context) {
+    throw new Error('TODO');
```

Comment:
I assume these could be used outside of Carousel..no ?

Reply:
I think only `Carousel.Dots` would be useful outside `Carousel`, so I've added some extra props to that.